auction/oracle.py

```diff
@@ -0,0 +1,272 @@
+                T, V_T, V_T_subset, Q, n_buyers, grades, f_hat)
+            border_cons.append(con)
+
+    elif problem_type == 'multi_good':
+        # we use the belloni algorithm separately for each object (grade) j
+        for j in grades:
+            Q_hat_ix = np.argsort(Q[j])[::-1]  # to get descending
+
+            E_Qs = []
+            for i in range(len(V_T)):
+                subset_ix = list(range(0, i + 1))
+                subset = Q_hat_ix[subset_ix]
+
+                E_Qs.append(subset.tolist())
+
+            V_T_subsets = E_Qs
+
+            for V_T_subset in V_T_subsets:
+                con = _check_one_border(
+                    T, V_T, V_T_subset, Q, n_buyers, [j], f_hat)
+                border_cons.append(con)
+
+    logger.debug('border constraint checks completed!')
+    end = time()
+    elapsed = (end - start)
+    n_violated, n_binding, n_inactive = 0, 0, 0
+    for c in border_cons:
+        if c.status == 'VIOLATED':
+            n_violated += 1
+        if c.status == 'BINDING':
+            n_binding += 1
+        if c.status == 'INACTIVE':
+            n_inactive += 1
+    logger.debug(
+        ('time taken (%s mins), # violated (%s), # binding (%s), '
+         '# inactive (%s)') %
+        (np.round(elapsed / 60, 1), n_violated, n_binding, n_inactive))
+
+    return border_cons
+
+
+def _check_one_border(T, V_T, V_T_subset, Q, n_buyers, grades, f_hat):
+    name = '%s_%s' % (
+        BORDER_PREFIX, '_'.join([str(i) for i in V_T_subset]))
+
+    # NOTE we `force_symmetric`=False here because we are using Q_vals so
+    # the indexing switch has already been done!
+    val = border_lhs_minus_rhs(
+        T, V_T, V_T_subset, Q, n_buyers, grades, f_hat, force_symmetric=False)
+
+    # NOTE becase we are comparing floating point numbers we eval
+    # `is_close` before we asses > / <
+    is_close = np.isclose(val, 0.0)
+    if is_close:
+        con = Constraint(name, None, 'BINDING')
+    elif val > 0:
+        con = Constraint(name, None, 'VIOLATED')
+    else:  # lhs_minus_rhs < 0:
+        con = Constraint(name, None, 'INACTIVE')
+
+    return con
+
+
+def separation_oracle(
+        Q, U, V_T, T, grades, n_buyers, f_hat, check_local_ic, problem_type,
+        net_size):
+    # TODO multiprocess...
+    logger.debug('starting separation oracle...')
+    I_ic, A_ic, B_ic = [], [], []
+    ic_cons = _check_ic(
+        Q, U, V_T, T, grades, check_local_ic, net_size)
+    for ic_con in ic_cons:
+        if ic_con.status == 'INACTIVE':
+            I_ic.append(ic_con)
+        if ic_con.status == 'VIOLATED':
+            A_ic.append(ic_con)
+        if ic_con.status == 'BINDING':
+            B_ic.append(ic_con)
+
+    I_border, A_border, B_border = [], [], []
+    # if N=1 skip Border checks
+    if n_buyers != 1:
+        border_cons = _check_border(
+            V_T, T, Q, grades, n_buyers, f_hat, problem_type)
+        for border_con in border_cons:
+            if border_con.status == 'INACTIVE':
```

Comment:
There is probably an opportunity to improve code reuse here also if you are ambitious. You can decompose and pass  ic_con and border_con as a variable e.g: 

```
def _check_status(constraint, I, A, B):
       if constaint.status == 'Inactive'
       .... 

```